hs/src/Reach/Connector/ETH_Solidity.hs

```diff
@@ -1357,15 +1361,22 @@ apiDef who ApiInfo {..} = do
                 ]
           tc <- solType_ $ vsToType ai_msg_vs
           let ty = solApply tc ["_vt"]
-          return $ (ty, argDefns, lifts, args)
+          return $ (ty, argDefns, lifts, args, m_arg_ty)
         AIC_SpreadArg -> do
           (args, argDefns) <-
             case ai_msg_tys of
               [T_Tuple ts] -> mkArgDefns ts
               _ -> impossible "apiDef: Expected one tuple arg"
           ty <- makeConsensusArg args
-          return $ (ty, argDefns, [], args)
-  (ty, argDefns, tyLifts, args) <- go ai_compile
+          return $ (ty, argDefns, [], args, m_arg_ty)
+  go ai_compile
+
+apiDef :: SLPart -> Bool -> ApiInfo -> App Doc
+apiDef who qualify ApiInfo {..} = do
+  let (prefix, suffix) = bool ("", "") ("_", show ai_which) qualify
+  let who_s = prefix <> bunpack who <> suffix
```

Comment:
Here's that pattern again... I think it's not something that is just in ALGO, but something that should be a shared DLBase resource